### DISCUSSION TITLE
Factory - try to create

### DISCUSSION THREAD

**Discussion by KrzysiekNDS** [laravel/framework#52948] [Ideas]
In Eloquent, when we use a factory to create a new record, it always tries to add it to the database. This can cause an error if a record with the same id already exists. Right now, we have to manually check if the record is there before creating it.

To make this easier, a new `tryToCreate` method could be added. This method would first check if a record with the given `id` is already in the database. If it exists, it wouldn’t create anything. If it doesn’t, it would create the record. This would be useful for tests where we just need to make sure a record with a specific `id` exists.

### Example
When a user with id 1 already exists in the database.
#### This will throw an QueryException:
```php
User::factory(['id' => 1])->create();
```
This should just try to create.
```php
// Using the new tryToCreate method
User::factory(['id' => 1])->tryToCreate();
```

If a user with `id` 1 already exists, nothing happens. If not, a new user with that `id` is created. This will save time by not needing to check for the record manually.

**Comment by murilolivorato** [comment]
```
<?php

namespace App\Factories;

use Illuminate\Database\Eloquent\Factories\Factory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\ModelNotFoundException;

class CustomFactory extends Factory
{
    /**
     * Try to create a new model instance.
     *
     * @param array $attributes
     * @return \Illuminate\Database\Eloquent\Model|void
     */
    public function tryToCreate(array $attributes = [])
    {
        $modelClass = $this->modelName();
        $primaryKey = (new $modelClass)->getKeyName();

        if (isset($attributes[$primaryKey])) {
            try {
                $modelClass::findOrFail($attributes[$primaryKey]);
                // Record already exists, do nothing
                return;
            } catch (ModelNotFoundException $e) {
                // Record does not exist, proceed to create
            }
        }

        return $this->create($attributes);
    }
}

```


```

<?php

namespace Database\Factories;

use App\Factories\CustomFactory;
use App\Models\User;
use Illuminate\Database\Eloquent\Factories\Factory;

class UserFactory extends CustomFactory
{
    protected $model = User::class;

    public function definition()
    {
        return [
            'name' => $this->faker->name,
            'email' => $this->faker->unique()->safeEmail,
            'password' => bcrypt('password'),
        ];
    }
}

```



```
<?php

use App\Models\User;
use Database\Factories\UserFactory;

public function testTryToCreate()
{
    // This will create a user with id 1 if it doesn't exist
    User::factory(['id' => 1])->tryToCreate();

    // Assert that the user with id 1 exists
    $this->assertDatabaseHas('users', ['id' => 1]);
}

```

**Comment by henzeb** [comment]
I think this is codesmell, you create a record for tests to make sure the user is in the correct state before the test runs.

suppose you want to test if a phone number is changed. If you would try create the user and the user exists, the phonenumber might be already changed as well. Someone changes the code, the code no longer updates the phone number, but your test still succeeds. you can circumvent that by either adding a precondition test, or you can truncate the table before you do that. adding a tryToCreate  and a precondition test is too much work for just using Model::truncate in your afterEach/teardown hook.
